### PR TITLE
Solidity v0.8.0

### LINKED ISSUE #168: [ETHEREUM-CONTRACTS] support solidity 0.8
Maybe, some day, v1.5 or later

## Notes

Notable features:

- revert with constant (enum-ish) instead of string. Can etherscan support it natively though?
- native SafeMaths support.
- user defined types

Caveats:

- **May** need to provide auto-generated compatibility interface files for 0.7.x

### PR DESCRIPTION
**Pull request by 0xdavinchee**
* payable(msg.sender).transfer applied where relevant
* intermediate conversions applied (**EXTRA REVIEW REQUIRED**)
* use >= 0.8.0 for interfaces
* use 0.8.12 for contracts
* OZ 4.5.0 from 3.4.0
* script for testing if solc 0.8.0 can compile header files
* a few test cases expected values changed
* natpsec comment cleanup
* typo cleanup

### CODE REVIEW

File: packages/ethereum-contracts/test/contracts/misc.test.js [hunk 1]
```diff
@@ -208,12 +208,12 @@ describe("Miscellaneous for test coverages", function () {
 
             await expectRevert(
                 tester.testInt96SafeMathDiv(MIN_INT96, 0),
-                "testInt96SafeMathDiv overflow"
+                "revert"
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/misc.test.js]: will find a solution to this under the OZ-tester framework

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 2]
```diff
@@ -104,8 +98,9 @@ contract ConstantFlowAgreementV1 is
         require(deposit < 2**95, "CFA: deposit number too big");
         deposit = _clipDepositNumberRoundingDown(deposit);
         (uint256 liquidationPeriod, ) = _decode3PsData(token);
-        uint256 flowrate1 = deposit.div(liquidationPeriod);
-        return int96(flowrate1);
+        uint256 flowrate1 = deposit / liquidationPeriod;
+        // REVIEW (0.8.12): - intermediate conversion step required
+        return int96(int256(flowrate1));
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:103]: In isolation it looks like it should use safecast for downcasting, but the the context showing that deposit is already constraint to `2**95`, hence safecast is not necessary.

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 3]
```diff
@@ -120,7 +115,9 @@ contract ConstantFlowAgreementV1 is
         uint256 minimumDeposit = gov.getConfigAsUint256(host, token, SUPERTOKEN_MINIMUM_DEPOSIT_KEY);
         uint256 pppConfig = gov.getConfigAsUint256(host, token, CFAV1_PPP_CONFIG_KEY);
         (uint256 liquidationPeriod, ) = SuperfluidGovernanceConfigs.decodePPPConfig(pppConfig);
-        require(uint256(flowRate).mul(liquidationPeriod) <= uint256(type(int96).max), "CFA: flow rate too big");
+        // REVIEW (0.8.12): - intermediate conversion step required
+        require(uint256(int256(flowRate))
+            * liquidationPeriod <= uint256(int256(type(int96).max)), "CFA: flow rate too big");
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:119]: it's okay.

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 4]
```diff
@@ -955,7 +953,8 @@ contract ConstantFlowAgreementV1 is
         exist = wordA > 0;
         if (exist) {
             flowData.timestamp = uint32(wordA >> 224);
-            flowData.flowRate = int96((wordA >> 128) & uint256(type(uint96).max));
+            // REVIEW (0.8.12): - intermediate conversion step required
+            flowData.flowRate = int96(int256(wordA >> 128) & int256(uint256(type(uint96).max)));
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:954]: mkay.

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol [hunk 5]
```diff
@@ -903,8 +899,10 @@ contract ConstantFlowAgreementV1 is
         returns(uint256 deposit)
     {
         if (flowRate == 0) return 0;
-        assert(liquidationPeriod <= uint256(type(int96).max));
-        deposit = uint256(flowRate.mul(int96(uint96(liquidationPeriod)), "CFA: deposit overflow"));
+        
+        // REVIEW (0.8.12): - intermediate conversion step required
+        assert(liquidationPeriod <= uint256(int256(type(int96).max)));
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:901]: mkay

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol (hunk 4 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:954]: I would add a comment that

"the second case is upcasting from type(uint96).max-> uint256->int256, here we do not need to use safecast"

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 6]
```diff
@@ -128,8 +122,9 @@ contract InstantDistributionAgreementV1 is
                 IndexData memory idata;
                 (exist, idata) = _getIndexData(token, iId);
                 assert(exist);
-                dynamicBalance = dynamicBalance.add(
-                    int256(idata.indexValue - sdata.indexValue) * int256(sdata.units)
+                dynamicBalance = dynamicBalance + (
+                    // REVIEW (0.8.12): - intermediate conversion step required
+                    int256(uint256(idata.indexValue - sdata.indexValue)) * int256(uint256(sdata.units))
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol:128]: for all reasonable "implicit" upcasting, let's add a comment to help readers to not to be surprised.

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 7]
```diff
@@ -270,10 +265,12 @@ contract InstantDistributionAgreementV1 is
         // - settle the publisher balance INSTANT-ly (ding ding ding, IDA)
         //   - adjust static balance directly
         token.settleBalance(publisher,
-            (-int256(newIndexValue - idata.indexValue)).mul(int256(idata.totalUnitsApproved)));
+            // REVIEW (0.8.12): - intermediate conversion step required
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: ditto

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 8]
```diff
@@ -270,10 +265,12 @@ contract InstantDistributionAgreementV1 is
...
+            (-int256(uint256(newIndexValue - idata.indexValue))) * int256(uint256(idata.totalUnitsApproved)));
         //   - adjust the publisher's deposit amount
         _adjustPublisherDeposit(token, publisher,
-            int256(newIndexValue - idata.indexValue).mul(int256(idata.totalUnitsPending)));
+            // REVIEW (0.8.12): - intermediate conversion step required
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: ditto

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 9]
```diff
@@ -384,8 +381,9 @@ contract InstantDistributionAgreementV1 is
         } else {
             cbStates.noopBit = SuperAppDefinitions.BEFORE_AGREEMENT_UPDATED_NOOP;
             vars.cbdata = AgreementLibrary.callAppBeforeCallback(cbStates, newCtx);
-
-            int balanceDelta = int256(vars.idata.indexValue - vars.sdata.indexValue) * int256(vars.sdata.units);
+            // REVIEW (0.8.12): - intermediate conversion step required
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: ditto

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 10]
```diff
@@ -449,8 +447,9 @@ contract InstantDistributionAgreementV1 is
 
         cbStates.noopBit = SuperAppDefinitions.BEFORE_AGREEMENT_TERMINATED_NOOP;
         vars.cbdata = AgreementLibrary.callAppBeforeCallback(cbStates, newCtx);
-
-        int256 balanceDelta = int256(vars.idata.indexValue - vars.sdata.indexValue) * int256(vars.sdata.units);
+        // REVIEW (0.8.12): - intermediate conversion step required
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: ditto

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 11]
```diff
@@ -556,8 +555,9 @@ contract InstantDistributionAgreementV1 is
             vars.idata.totalUnitsPending = vars.idata.totalUnitsPending.add(units, "IDA: E_OVERFLOW");
             token.updateAgreementData(vars.iId, _encodeIndexData(vars.idata));
         }
-
-        int256 balanceDelta = int256(vars.idata.indexValue - vars.sdata.indexValue) * int256(vars.sdata.units);
+        // REVIEW (0.8.12): - intermediate conversion step required
+        int256 balanceDelta = int256(uint256(vars.idata.indexValue - vars.sdata.indexValue))
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol:565]: ditto

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 12]
```diff
@@ -728,8 +728,9 @@ contract InstantDistributionAgreementV1 is
 
         cbStates.noopBit = SuperAppDefinitions.BEFORE_AGREEMENT_TERMINATED_NOOP;
         vars.cbdata = AgreementLibrary.callAppBeforeCallback(cbStates, newCtx);
-
-        int256 balanceDelta = int256(vars.idata.indexValue - vars.sdata.indexValue) * int256(vars.sdata.units);
+        // REVIEW (0.8.12): - intermediate conversion step required
+        int256 balanceDelta = int256(uint256(vars.idata.indexValue - vars.sdata.indexValue))
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol:739]: ditto

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 13]
```diff
@@ -926,8 +927,10 @@ contract InstantDistributionAgreementV1 is
         uint256 b = uint256(adata[1]);
         exist = a > 0;
         if (exist) {
-            idata.indexValue = uint128(a & uint256(int128(-1)));
-            idata.totalUnitsApproved = uint128(b & uint256(int128(-1)));
+            // REVIEW (0.8.12): - negating uint
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: The code doesn't make sense.

To solve this once and for all, let's just prove that the brutal version of unsafe downcast is okay to be used.

```
// SPDX-License-Identifier: GPL-3.0-or-later
pragma solidity ^0.8.6;

contract TestUnsafeDowncast {

    function unsafeDowncast1(uint256 u) pure external returns (uint128) {
        return uint128(u);
    }

    function unsafeDowncast2(uint256 u) pure external returns (uint128) {
        return uint128(u & uint256(type(uint128).max));
    }

    function unsafeDowncast3(uint256 u) pure external returns (uint128) {
        return uint128(u & type(uint256).max - 1 + 1);
    }
}
```

```
// SPDX-License-Identifier: GPL-3.0-or-later
pragma solidity ^0.8.6;

import "ds-test/test.sol";

import "./TestUnsafeDowncast.sol";

contract TestUnsafeDowncastTest is DSTest {
    TestUnsafeDowncast downcast;

    function setUp() public {
        downcast = new TestUnsafeDowncast();
    }

    function prove_1eq2(uint256 v) public {
        assertEq(downcast.unsafeDowncast1(v), downcast.unsafeDowncast2(v));
    }

    function prove_1eq3(uint256 v) public {
        assertEq(downcast.unsafeDowncast1(v), downcast.unsafeDowncast3(v));
    }
}

```


provable that we should just use the unsafeDowncast brutal version, and add bunch of comments about it

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 14]
```diff
@@ -965,7 +968,8 @@ contract InstantDistributionAgreementV1 is
             publisher,
             _PUBLISHER_DEPOSIT_STATE_SLOT_ID,
             1);
-        data[0] = bytes32(int256(data[0]) + delta);
+        // REVIEW (0.8.12): - intermediate conversion step required
+        data[0] = bytes32(uint256(int256(uint256(data[0])) + delta));
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: here we should use SafeCast.toInt256()

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 15]
```diff
@@ -989,7 +993,8 @@ contract InstantDistributionAgreementV1 is
     {
         data = new bytes32[](2);
         data[0] = bytes32(
-            (uint256(sdata.publisher) << (12*8)) |
+            // REVIEW (0.8.12): - intermediate conversion step required
+            (uint256(uint160(sdata.publisher)) << (12*8)) |
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol:1003]: this is upCast, so it should be okay

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol [hunk 16]
```diff
@@ -1014,7 +1019,8 @@ contract InstantDistributionAgreementV1 is
             sdata.publisher = address(uint160(a >> (12*8)));
             sdata.indexId = uint32((a >> 32) & type(uint32).max);
             sdata.subId = uint32(a & type(uint32).max);
-            sdata.indexValue = uint128(b & uint256(int128(-1)));
+            // REVIEW (0.8.12): - negating uint
+            sdata.indexValue = uint128(b & type(uint256).max - 1 + 1);
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: similar as one of the previous cases -> use unsafeDowncast and add comments

File: packages/ethereum-contracts/contracts/interfaces/superfluid/SuperAppBase.sol [hunk 17]
```diff
@@ -1,12 +1,12 @@
 // SPDX-License-Identifier: AGPLv3
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/SuperAppBase.sol]: let's move this back

File: packages/ethereum-contracts/contracts/libs/Strings.sol [hunk 18]
```diff
@@ -215,7 +215,9 @@ library Strings {
             }
             if (a != b) {
                 // Mask out irrelevant bytes and check again
-                uint256 mask = uint256(-1); // 0xffff...
+                // REVIEW (0.8.12): negating uint
+                // uint256 mask = uint256(-1); // 0xffff...
+                uint256 mask = type(uint256).max - 1 + 1; // 0xffff...
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/libs/Strings.sol]: use type(uint256).max

File: packages/ethereum-contracts/contracts/mocks/AgreementMock.sol [hunk 19]
```diff
@@ -48,7 +46,8 @@ contract AgreementMock is AgreementBase {
         bytes32[] memory slotData = token.getAgreementStateSlot(
             address(this), account, _REAL_TIME_BALANCE_SLOT_ID, 3);
         return (
-            int256(slotData[0]),
+            // REVIEW (0.8.12): - intermediate conversion step required
+            int256(uint256(slotData[0])),
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/mocks/AgreementMock.sol]: use SafeCast

File: packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol [hunk 20]
```diff
@@ -143,7 +143,8 @@ abstract contract SuperTokenFactoryBase is
         external override
     {
         // odd solidity stuff..
-        address payable a = address(uint160(customSuperTokenProxy));
+        // REVIEW (0.8.12): - intermediate conversion step required
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol]: mkay

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js [hunk 21]
```diff
@@ -375,7 +375,7 @@ describe("Using ConstantFlowAgreement v1", function () {
                         receiver: t.aliases.carol,
                         flowRate: MAXIMUM_FLOW_RATE.toString(),
                     }),
-                    "CFA: deposit overflow"
+                    "CallUtils: target reverted"
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js]: Let's catch the custom error

File: packages/ethereum-contracts/test/contracts/misc.test.js [hunk 22]
```diff
@@ -122,7 +122,7 @@ describe("Miscellaneous for test coverages", function () {
             );
             await expectRevert(
                 tester.testInt96SafeMathMul(MAX_INT96_DIV_2_PLUS_1, 2),
-                "testInt96SafeMathMul overflow"
+                "revert"
```
Comment (hellwolf) [inline packages/ethereum-contracts/test/contracts/misc.test.js]: let's fix it

**Review by hellwolf — COMMENTED** [review]
good stuff. left some comments and requested changes.

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol (hunk 2 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:103]:
```
// NOTE downcasting is safe as we constrain deposit to less than
// 2 ** 95 so the resulting value flowRate1 will fit into int96
```

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol (hunk 3 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:119]: upcasting is ok

File: packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol (hunk 5 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:901]: upcasting ok

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol (hunk 13 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: unsafe downcast gist:
https://gist.github.com/0xdavinchee/9834dc689543f19ec07872ad7d766b09

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol (hunk 14 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: safecasting here now

File: packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol (hunk 16 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol]: documented

File: packages/ethereum-contracts/contracts/interfaces/superfluid/SuperAppBase.sol (hunk 17 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/SuperAppBase.sol]: superAppBase moved back to `contracts/apps`

File: packages/ethereum-contracts/contracts/libs/Strings.sol (hunk 18 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/libs/Strings.sol]: resolved

File: packages/ethereum-contracts/contracts/mocks/AgreementMock.sol (hunk 19 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/mocks/AgreementMock.sol]: using safe cast now

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js (hunk 21 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js]: this is caught with the unchecked fix

File: packages/ethereum-contracts/test/contracts/misc.test.js (hunk 22 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/test/contracts/misc.test.js]: caught w/ unchecked fix

**Review by hellwolf — APPROVED** [review]
We will have cookies and eat it.

image

**Pull request merged**